### ISSUE TITLE
Transaction details missing on edit view

### ISSUE THREAD

**Issue by psychowood** [firefly-iii/firefly-iii#3704]
**Bug description**
I am running Firefly III v5.4.0-alpha.1, and my problem is:

Transaction edit view ( /transactions/edit/XXXXX ) sometimes just shows the bottom (submit) panel:

<img width="950" alt="image" src="https://user-images.githubusercontent.com/115389/90331734-228e9400-dfb7-11ea-8770-6971a7d5cbaa.png">

**Steps to reproduce**
I'm not 100% sure. It seems that if I Cmd+Tab the same url on another tab, it works. 
Edit button from firefly-iii, Cmd+R or send Enter in the address bar, it does not reload correctly.

I'll send you an email with the two .har files for comparing the sessions when I managed to replicate it on the demo site (I'm not sure there may be tracking cookies, I'd avoid posting publicly), the main differences I noticed are:

- Missing currencies API call in the KO sample
- "pageref" is always "page_2" in the KO sample (this could be related to har saving).

**Bonus points**
I was able to replicate it on the demo site https://demo.firefly-iii.org/ but it seems to be not as frequent as on my system.

---

Debug information generated at 2020-08-16 12:40:45 Europe/Rome for Firefly III version **5.4.0-alpha.1**.

| Scope | Version |
| --- | --- |
| Firefly III | 5.4.0-alpha.1 |
| Firefly III API | 1.4.0 |
| PHP | 7.4.8 |
| Host | Linux |

| System info | Value |
| --- | --- |
| Installation ID | 3c9524c2-6757-4aec-8d85-5387234150a5 |
| Using docker? | true |
| Telemetry | false |
| Layout | v1 |
| App environment | local |
| App debug mode | false |
| App cache driver | file |
| App logging | debug, stack |
| Display errors | Off |
| Error reporting | ALL errors |
| Interface | apache2handler |
| Default language | it_IT |
| Default locale | it_IT |
| BCscale | 12 |
| DB drivers | mysql, pgsql, sqlite |
| Current driver | mysql |
| Login provider | eloquent  |
| Trusted proxies (.env) | ** |

| User info | Value |
| --- | --- |
| Session start | 2020-08-01 00:00:00 |
| Session end | 2020-08-31 23:59:59 |
| Session first | 2008-02-08 00:00:00 |
| User ID | 1 |
| User language | en_US |
| User locale | en_US |
| Attempt at "en_US.utf8" | 'en_US.utf8' |
| Attempt at "en_US.UTF-8" | 'en_US.UTF-8' |
| User agent | Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:79.0) Gecko/20100101 Firefox/79.0 |

---

```
Truncated from this point <----|tables.  
[2020-08-16 12:05:15] local.INFO: Configured DB version (15) equals expected DB version (15)  
[2020-08-16 12:05:15] local.INFO: Installed Firefly III version (5.4.0-alpha.1) equals expected Firefly III version (5.4.0-alpha.1)  
[2020-08-16 12:05:15] local.DEBUG: Will NOT set previous URL to http://dockervm:8081/login  
[2020-08-16 12:05:15] local.DEBUG: Will NOT set previous URL to http://dockervm:8081/login  
[2020-08-16 12:05:19] local.DEBUG: Will NOT set previous URL to http://dockervm:8081/v1/jscript/variables?ext=.js&v=5.4.0-alpha.1  
[2020-08-16 12:05:20] local.DEBUG: GroupCollector::startQuery  
[2020-08-16 12:05:20] local.DEBUG: GroupCollector::startQuery  
[2020-08-16 12:07:56] local.DEBUG: Going to try to decrypt users currency preference.  
[2020-08-16 12:07:56] local.DEBUG: Could not decrypt "EUR". The payload is invalid.  
[2020-08-16 12:07:56] local.DEBUG: Searching for "Cash account" of type "Cash account"  
[2020-08-16 12:07:56] local.DEBUG: Saving URL http://dockervm:8081/transactions/edit/10310 under key transactions.edit.uri  
[2020-08-16 12:07:56] local.DEBUG: Will set previous URL to http://dockervm:8081/transactions/edit/10310  
[2020-08-16 12:07:59] local.DEBUG: Will NOT set previous URL to http://dockervm:8081/v1/jscript/variables?ext=.js&v=5.4.0-alpha.1  
[2020-08-16 12:08:00] local.DEBUG: GroupCollector::startQuery  
[2020-08-16 12:08:00] local.DEBUG: GroupCollector::startQuery  
[2020-08-16 12:08:07] local.DEBUG: Going to try to decrypt users currency preference.  
[2020-08-16 12:08:07] local.DEBUG: Could not decrypt "EUR". The payload is invalid.  
[2020-08-16 12:08:07] local.DEBUG: Searching for "Cash account" of type "Cash account"  
[2020-08-16 12:08:07] local.DEBUG: Saving URL http://dockervm:8081/transactions/edit/10310 under key transactions.edit.uri  
[2020-08-16 12:08:07] local.DEBUG: Will set previous URL to http://dockervm:8081/transactions/edit/10310  
[2020-08-16 12:08:07] local.DEBUG: Will NOT set previous URL to http://dockervm:8081/v1/jscript/variables?ext=.js&v=5.4.0-alpha.1  
[2020-08-16 12:08:08] local.DEBUG: GroupCollector::startQuery  
[2020-08-16 12:08:08] local.DEBUG: GroupCollector::startQuery  
[2020-08-16 12:09:07] local.DEBUG: Will set previous URL to http://dockervm:8081/transactions/edit/10310  
[2020-08-16 12:09:07] local.DEBUG: Will NOT set previous URL to http://dockervm:8081/v1/jscript/variables?ext=.js&v=5.4.0-alpha.1  
[2020-08-16 12:09:07] local.DEBUG: Installer middleware for URI http://dockervm:8081/login  
[2020-08-16 12:09:07] local.DEBUG: Now in routine hasNoTables()  
[2020-08-16 12:09:07] local.DEBUG: Installer middleware for URI http://dockervm:8081/login  
[2020-08-16 12:09:07] local.DEBUG: Now in routine hasNoTables()  
[2020-08-16 12:09:07] local.DEBUG: Everything seems OK with the tables.  
[2020-08-16 12:09:07] local.DEBUG: Everything seems OK with the tables.  
[2020-08-16 12:09:07] local.INFO: Configured DB version (15) equals expected DB version (15)  
[2020-08-16 12:09:07] local.INFO: Installed Firefly III version (5.4.0-alpha.1) equals expected Firefly III version (5.4.0-alpha.1)  
[2020-08-16 12:09:07] local.INFO: Configured DB version (15) equals expected DB version (15)  
[2020-08-16 12:09:07] local.INFO: Installed Firefly III version (5.4.0-alpha.1) equals expected Firefly III version (5.4.0-alpha.1)  
[2020-08-16 12:09:07] local.DEBUG: Will NOT set previous URL to http://dockervm:8081/login  
[2020-08-16 12:09:07] local.DEBUG: Will NOT set previous URL to http://dockervm:8081/login  
[2020-08-16 12:09:10] local.DEBUG: Going to try to decrypt users currency preference.  
[2020-08-16 12:09:10] local.DEBUG: Could not decrypt "EUR". The payload is invalid.  
[2020-08-16 12:09:10] local.DEBUG: Searching for "Cash account" of type "Cash account"  
[2020-08-16 12:09:10] local.DEBUG: Saving URL http://dockervm:8081/transactions/edit/10310 under key transactions.edit.uri  
[2020-08-16 12:09:10] local.DEBUG: Will set previous URL to http://dockervm:8081/transactions/edit/10310  
[2020-08-16 12:09:12] local.DEBUG: Will NOT set previous URL to http://dockervm:8081/v1/jscript/variables?ext=.js&v=5.4.0-alpha.1  
[2020-08-16 12:09:14] local.DEBUG: GroupCollector::startQuery  
[2020-08-16 12:09:14] local.DEBUG: GroupCollector::startQuery  
[2020-08-16 12:09:19] local.DEBUG: Going to try to decrypt users currency preference.  
[2020-08-16 12:09:19] local.DEBUG: Could not decrypt "EUR". The payload is invalid.  
[2020-08-16 12:09:19] local.DEBUG: Searching for "Cash account" of type "Cash account"  
[2020-08-16 12:09:19] local.DEBUG: Saving URL http://dockervm:8081/transactions/edit/10310 under key transactions.edit.uri  
[2020-08-16 12:09:19] local.DEBUG: Will set previous URL to http://dockervm:8081/transactions/edit/10310  
[2020-08-16 12:09:20] local.DEBUG: Will NOT set previous URL to http://dockervm:8081/v1/jscript/variables?ext=.js&v=5.4.0-alpha.1  
[2020-08-16 12:09:22] local.DEBUG: GroupCollector::startQuery  
[2020-08-16 12:09:22] local.DEBUG: GroupCollector::startQuery  
[2020-08-16 12:09:25] local.DEBUG: Going to try to decrypt users currency preference.  
[2020-08-16 12:09:25] local.DEBUG: Could not decrypt "EUR". The payload is invalid.  
[2020-08-16 12:09:25] local.DEBUG: Searching for "Cash account" of type "Cash account"  
[2020-08-16 12:09:25] local.DEBUG: Saving URL http://dockervm:8081/transactions/edit/10310 under key transactions.edit.uri  
[2020-08-16 12:09:25] local.DEBUG: Will set previous URL to http://dockervm:8081/transactions/edit/10310  
[2020-08-16 12:09:26] local.DEBUG: Will NOT set previous URL to http://dockervm:8081/v1/jscript/variables?ext=.js&v=5.4.0-alpha.1  
[2020-08-16 12:09:26] local.DEBUG: GroupCollector::startQuery  
[2020-08-16 12:09:26] local.DEBUG: GroupCollector::startQuery  
[2020-08-16 12:09:32] local.DEBUG: Going to try to decrypt users currency preference.  
[2020-08-16 12:09:32] local.DEBUG: Could not decrypt "EUR". The payload is invalid.  
[2020-08-16 12:09:32] local.DEBUG: Searching for "Cash account" of type "Cash account"  
[2020-08-16 12:09:32] local.DEBUG: Saving URL http://dockervm:8081/transactions/edit/10310 under key transactions.edit.uri  
[2020-08-16 12:09:32] local.DEBUG: Will set previous URL to http://dockervm:8081/transactions/edit/10310  
[2020-08-16 12:09:33] local.DEBUG: Will NOT set previous URL to http://dockervm:8081/v1/jscript/variables?ext=.js&v=5.4.0-alpha.1  
[2020-08-16 12:09:33] local.DEBUG: GroupCollector::startQuery  
[2020-08-16 12:09:33] local.DEBUG: GroupCollector::startQuery  
[2020-08-16 12:40:25] local.DEBUG: Trying to set en_US.utf8  
[2020-08-16 12:40:25] local.DEBUG: Trying to set en_US.UTF-8  
[2020-08-16 12:40:25] local.DEBUG: Will set previous URL to http://dockervm:8081/debug  
[2020-08-16 12:40:40] local.DEBUG: Going to try to decrypt users currency preference.  
[2020-08-16 12:40:40] local.DEBUG: Could not decrypt "EUR". The payload is invalid.  
[2020-08-16 12:40:40] local.DEBUG: Searching for "Cash account" of type "Cash account"  
[2020-08-16 12:40:40] local.DEBUG: Will set previous URL to http://dockervm:8081/transactions/edit/6009  
[2020-08-16 12:40:42] local.DEBUG: Will NOT set previous URL to http://dockervm:8081/v1/jscript/variables?ext=.js&v=5.4.0-alpha.1  
[2020-08-16 12:40:43] local.DEBUG: GroupCollector::startQuery  
[2020-08-16 12:40:43] local.DEBUG: GroupCollector::startQuery  
[2020-08-16 12:40:45] local.DEBUG: Trying to set en_US.utf8  
[2020-08-16 12:40:45] local.DEBUG: Trying to set en_US.UTF-8  

```

**Comment by JC5** [comment]
> "pageref" is always "page_2"


I don’t use these parameters. I can’t do a deep dive until after the weekend.

**Comment by JC5** [comment]
I see nothing special in your har files, I'm afraid.

**Comment by psychowood** [comment]
Ok thanks, I'll try to investigate.

**Comment by psychowood** [comment]
TL;DR

It happens on Firefox, both on macOS and Android, even in private mode. While trying the vue-devtools I messed with security settings on firefox, and in the end completely disabling CSP (security.csp.enable = false) seems to have solved the issue.

The fact that it does not happen every time makes me think it could be caused by some race conditions, caching and so on: do you set/edit CSP headers on some XHR calls?

---

I was able to replicate it on:

- Firefox (v79.0) for macOS
- Firefox (v68.11.0) for Android
- Firefox Nightly (build 7/28 @ 6.29 AM) for Android

It happens regardless of enabled addons, even in private mode.

No problems on Chrome/Safari/Samsung Internet

Steps to replicate it consistently:
- Open Firefox
- Go directly to https://demo.firefly-iii.org/transactions/edit/303
- Login with demo@firefly:demo (leaving "remember me" unchecked, should not be relevant tho)
- Wait for the page load
- "Success":

<img width="390" alt="image" src="https://user-images.githubusercontent.com/115389/90425966-e0517980-e0c0-11ea-83ee-72571a571da9.png">

Reload page (cmd+R) keeps the wrong behaviour, sometimes (only on macOS) sending "Enter" on the address bar loads the page correctly.

Thing I noticed:

- When the page does not load correctly, the console stops here:

<img width="688" alt="image" src="https://user-images.githubusercontent.com/115389/90428359-2825d000-e0c4-11ea-8f31-a871fb8816f2.png">

- When the page does load correctly, I see currencies logged:

<img width="690" alt="image" src="https://user-images.githubusercontent.com/115389/90428529-6b803e80-e0c4-11ea-9f0c-9ad04a78279a.png">

The main point is this: while trying to debug the demo site with vue-devtools, I notice the devtools did not recognize Vue correctly ("Vue.js not detected") showing this message:

<img width="715" alt="image" src="https://user-images.githubusercontent.com/115389/90430015-00843700-e0c7-11ea-9619-792b62bb73d2.png">

I then overrode CSP settings (about:config -> security.csp.enable = false), vue-devtools loaded correctly, and so the page... removing vue-devtools and keeping CSP disabled kept the site loading correctly every time.

**Comment by JC5** [comment]
I know vue has a hard time handling CSP. Thanks for checking it out. I’ll try to find some time.

**Comment by JC5** [comment]
This starts happening to me the moment I installed vue devtools.

**Comment by psychowood** [comment]
I noticed the explicit "Content Security Policy" when installing vue-tools too, but that was only the input to check for CSP settings.

If you can't replicate it on you Firefox is probably something related to my setups, I'll try on other platforms and let you know.

I'd say you wasted enough of your time on this, as long as I'm the only one reporting the issue...

**Comment by JC5** [comment]
Haha no worries. But I'm inclined to blame vue tools and leave you to it. If you have a lead let me know, I don't mind "wasting" time on Firefly III ;)

**Comment by sitenzo** [comment]
Hello,
Is there any progress on this issue?
I have a simular issue on version 5.3.3

I get a 401 on https://{{HIDDEN}}.nl/api/v1/transactions/25?_token={{HIDDEN}}

The logs shows this in pageload / pagereload..
[2020-08-31 21:18:18] local.DEBUG: Going to try to decrypt users currency preference.  
[2020-08-31 21:18:18] local.DEBUG: Could not decrypt "EUR". The payload is invalid.  
[2020-08-31 21:18:18] local.DEBUG: Searching for "Cash account" of type "Cash account"  
[2020-08-31 21:18:18] local.DEBUG: Saving URL https://{{HIDDEN}}/transactions/edit/25 under key transactions.edit.uri  
[2020-08-31 21:18:18] local.DEBUG: Will set previous URL to https://{{HIDDEN}}/transactions/edit/25  
[2020-08-31 21:18:18] local.DEBUG: Will NOT set previous URL to https://{{HIDDEN}}/v1/jscript/variables?ext=.js&v=5.3.3  

I can use all the other funtions.

**Comment by JC5** [comment]
Are you running vue dev tools?

**Comment by psychowood** [comment]
@JC5 just for clarification, I **don't** have vuetools installed, I tried to install it to debug the error and that lead me to discovering the CSP correlation.

Currently, with no addons installed, in private mode, opening directly [this page](https://demo.firefly-iii.org/transactions/edit/303) results in this:

<img width="391" alt="image" src="https://user-images.githubusercontent.com/115389/91852439-2251f200-ec61-11ea-8b3a-1063bdf5e4d5.png">

This is happening even with the latest Firefox for Android ("firefox daylight", rewritten from scratch they say) installed, that does not support addons at all.

Is there a way to enable more logging on the frontend, too?

@sitenzo can you share your https://{{HIDDEN}}/debug output?

**Comment by psychowood** [comment]
Agreed :) 
Personally I didn't have the issue on v5.3.3, it started with v5.4.0-alpha.1, I'm just trying it periodically.

**Comment by sitenzo** [comment]
@psychowood 

Debug information generated at 2020-09-01 18:52:32 Europe/Amsterdam for Firefly III version **5.3.3**.

| Scope | Version |
| --- | --- |
| Firefly III | 5.3.3 |
| Firefly III API | 1.3.0 |
| PHP | 7.4.9 |
| Host | Linux |

| System info | Value |
| --- | --- |
| Installation ID | afaf3a12-6a2b-4d03-b338-540f5a3b779b |
| Using docker? | false |
| Telemetry | false |
| Layout | v1 |
| App environment | local |
| App debug mode | false |
| App cache driver | file |
| App logging | warning, stack |
| Display errors | Off |
| Error reporting | ALL errors |
| Interface | litespeed |
| Default language | nl_NL |
| Default locale | equal |
| BCscale | 12 |
| DB drivers | mysql, sqlite |
| Current driver | mysql |
| Login provider | eloquent  |
| Trusted proxies (.env) |  |

| User info | Value |
| --- | --- |
| Session start | 2020-08-01 00:00:00 |
| Session end | 2020-08-31 00:00:00 |
| Session first | 2020-08-23 00:00:00 |
| User ID | 1 |
| User language | nl_NL |
| User locale | nl_NL |
| Attempt at "nl_NL.utf8" | 'nl_NL.utf8' |
| Attempt at "nl_NL.UTF-8" | 'nl_NL.UTF-8' |
| User agent | Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.135 Safari/537.36 |

**Comment by JC5** [comment]
>  I have a simular issue on version 5.3.3

I still have no idea what issue you have and you haven't answered my question if you happen to be running vue dev tools. Can we start with that please?

**Comment by sitenzo** [comment]
> I still have no idea what issue you have and you haven't answered my question if you happen to be running vue dev tools. Can we start with that please?

I can't edit transactions.
Not running veu dev tools
Not running anything else than the base Firefly.
Console.log shows a 401 forbidden error on /api/v1/transactions/25?_token=

![image](https://user-images.githubusercontent.com/52133032/91889670-7de6a480-ec8e-11ea-9edc-65fae0b8c454.png)

**Comment by JC5** [comment]
Thanks. No clue though. Let's start with the basics first: clear cookies, login again, clear cache, etc. Create transaction works fine? Do the auto-complete dropdowns on the create transaction form work?

**Comment by sitenzo** [comment]
All the auto-complete drop-dows work fine on transaction create page.
Import from csv-importer works.
Creating a transaction manually i get a "Fout! unauthenticated" (cleared cache, even used browser stack to check)

**Comment by JC5** [comment]
I can't replicate this. The dropdowns use the exact same API calls so I'm a bit lost. What do you mean with, creating a transaction manually?

**Comment by sitenzo** [comment]
i make use of csv-import instead of manually create a transaction.

This are the XHR on create transaction :

![image](https://user-images.githubusercontent.com/52133032/92307310-d8854680-ef95-11ea-9cdc-11a5277e386b.png)

This are the XHR calls on edit transaction : 

![image](https://user-images.githubusercontent.com/52133032/92307336-08cce500-ef96-11ea-8944-6adee10361fc.png)

It looks like some permission error.

**Comment by psychowood** [comment]
Some questions:
- are you experiencing it with different browsers or only with Firefox?
- the demo site is working correctly?
- could you try with another user?

I'm quite sure we are talking about a different issue than the one I opened, in my case I don't get 401 errors, and v5.3.3 is working.

**Comment by JC5** [comment]
picture-in-picture js is not mine. Not sure what that is.

**Comment by sitenzo** [comment]
@psychowood @JC5 
My problem was solved after reinstalling the self host version.
Sorry for the inconvenience.

**Comment by JC5** [comment]
No problem! Glad to hear that it works!
